smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -57,14 +73,97 @@ public ApplicationProtocol getApplicationProtocol() {
     @Override
     public void generateRequestSerializers(GenerationContext context) {
         TopDownIndex topDownIndex = context.getModel().getKnowledge(TopDownIndex.class);
+        OperationIndex operationIndex = context.getModel().getKnowledge(OperationIndex.class);
+        Walker shapeWalker = new Walker(context.getModel().getKnowledge(NeighborProviderIndex.class).getProvider());
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        TypeScriptWriter writer = context.getWriter();
+
+        // Track the shapes we need to generate sub-serializers for.
+        Set<Shape> serializingShapes = new HashSet<>();
         for (OperationShape operation : topDownIndex.getContainedOperations(context.getService())) {
             OptionalUtils.ifPresentOrElse(
                     operation.getTrait(HttpTrait.class),
                     httpTrait -> generateOperationSerializer(context, operation, httpTrait),
                     () -> LOGGER.warning(String.format(
                             "Unable to generate %s protocol request bindings for %s because it does not have an "
                             + "http binding trait", getName(), operation.getId())));
+
+            operationIndex.getInput(operation)
+                    .ifPresent(input -> serializingShapes.addAll(shapeWalker.walkShapes(input).stream()
+                            // Don't generate a sub-serializer for the actual input shape.
+                            .filter(s -> !input.equals(s))
+                            .collect(Collectors.toSet())));
         }
+
+        // TODO Abstraction point
+        // This should almost certainly be abstracted out to something else, because any protocol
+        // with a document body is going to need to generate sub-serializers.
+        // Generate the serializers for shapes within the operation closure.
+        serializingShapes.forEach(shape -> shape.accept(new ShapeVisitor.Default<Void>() {
```

Comment:
This anonymous visit is a lot to put into a single method. Is this something that can be easily moved into an inner class (it could be non-static if that would be useful).

Reply:
Yeah, I think that's part of getting this abstraction right. Unsure exactly what shape it's going to take so wanted to leave that to the author, even if that's me.